### PR TITLE
ceph: add option to set FSGroupPolicy for CSI PVC

### LINKED ISSUE #6830: Be able to use the k8s 1.20 fsGroupPolicy option via rook options
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Kubernetes 1.20 allows disabling files chown on request from CSI drivers. Rook has to configure this option in the CSI driver daemonset.

https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/

**What is use case behind this feature:**
Fixes https://github.com/rook/rook/issues/2254

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**
in kubernetes 1.20 it added support to set FSGroupPolicy in the CSI driver object, the same supported is added to both
CephFS and RBD. where the user will have an option to set the FSGroupPolicy for CephFS and RBD separately.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

**Which issue is resolved by this Pull Request:**
Resolves #6830 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by Madhu-1** [comment]
Note: To make use of this new setting the user needs to delete the CSI driver objects which are created earlier.

**Review by satoru-takeuchi — APPROVED** [review]

File: pkg/operator/ceph/csi/spec.go [hunk 1]
```diff
@@ -601,6 +611,12 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
 	}
 
 	if apierrors.IsNotFound(err) {
+		if fsGroupPolicy != "" {
+			policy := k8scsi.FSGroupPolicy(fsGroupPolicy)
+			// FsGroupPolicy field is immutable, should be set only during
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:616]: So if they change this policy they will need to delete the CSI driver and restart the operator? How about a line about this in the csi doc ceph-csi-drivers.md at the same time as adding the link to the helm settings?

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 2]
```diff
@@ -121,6 +121,18 @@ data:
   # See the upgrade guide: https://rook.io/docs/rook/master/ceph-upgrade.html
   CSI_FORCE_CEPHFS_KERNEL_CLIENT: "true"
 
+  # (Optional) policy to modifying a volume's ownership or permissions when the RBD PVC is being mounted.
+  # supported values are None, File, ReadWriteOnceWithFSType. if not specified nothing will be set when
+  # creating a CSIDriver object.
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:126]: If nothing is set, is it the same as one of these settings? If so, how about if we uncomment the setting here in the examples and set it to the default behavior? Then it's more clear what the default behavior is.

File: Documentation/helm-operator.md [hunk 3]
```diff
@@ -101,7 +101,9 @@ The following tables lists the configurable parameters of the rook-operator char
 | `csi.enableCephfsDriver`           | Enable Ceph CSI CephFS driver.                                                                                                 | `true`                                                    |
 | `csi.pluginPriorityClassName`      | PriorityClassName to be set on csi driver plugin pods.                                                                         | <none>                                                    |
 | `csi.provisionerPriorityClassName` | PriorityClassName to be set on csi driver provisioner pods.                                                                    | <none>                                                    |
-| `csi.enableOMAPGenerator`          | EnableOMAP generator deploys omap sidecar in CSI provisioner pod, to enable it set it to true                                  | `false`                                                    |
+| `csi.enableOMAPGenerator`          | EnableOMAP generator deploys omap sidecar in CSI provisioner pod, to enable it set it to true                                  | `false`                                                   |
```
Comment (travisn) [inline Documentation/helm-operator.md:104]: Since we only document the csi settings in the helm doc, there are a lot of csi settings that may be interesting to the user and they might not find them. How about adding a link from the ceph-csi-drivers.md to this helm doc where they can find the settings? Thought it could be a separate PR if you want.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go [hunk 4]
```diff
@@ -486,13 +486,23 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
 	}
 
 	if EnableRBD {
-		err = createCSIDriverInfo(ctx, clientset, RBDDriverName)
+		rbdfsGroupPolicy, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_RBD_FSGROUPPOLICY", "")
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:489]: Nit: Go syntax guidelines suggest this should be `rbdFsGroupPolicy` or `rbdFSGroupPolicy`

File: pkg/operator/ceph/csi/spec.go [hunk 5]
```diff
@@ -486,13 +486,23 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
...
+			logger.Warningf("failed to load CSI_RBD_FSGROUPPOLICY. %v", err)
+		}
+		err = createCSIDriverInfo(ctx, clientset, RBDDriverName, rbdfsGroupPolicy)
 		if err != nil {
 			return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
 		}
 	}
 	if EnableCephFS {
-		err = createCSIDriverInfo(ctx, clientset, CephFSDriverName)
+		cephFSfsGroupPolicy, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_CEPHFS_FSGROUPPOLICY", "")
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:500]: Nit: Go syntax guidelines suggest this should be `cephFSFsGroupPolicy`, `cephFsFsGroupPolicy`, or `cephFSFSGroupPolicy`. Trends below suggest cephFSFSGroupPolicy which is hard to read. 

`fsGroupPolicyForCephFS` is longer but avoids the weirdness.

Another option might be to create a `CSIDriverConfig` struct that could hold the driver name and fsGroupPolicy configs and allow for future config expansion without altering the `createCSIDriverInfo` function signatures. But this is definitely not necessary for just a couple configs; just an option.

File: pkg/operator/ceph/csi/spec.go (hunk 1 above)
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:616]:
```suggestion
			// FSGroupPolicy field is immutable, should be set only during
```

File: cluster/examples/kubernetes/ceph/operator.yaml [hunk 6]
```diff
@@ -45,6 +45,18 @@ data:
...
   CSI_FORCE_CEPHFS_KERNEL_CLIENT: "true"
 
+  # (Optional) policy to modifying a volume's ownership or permissions when the RBD PVC is being mounted.
+  # supported values are None, File, ReadWriteOnceWithFSType. if not specified nothing will be set when
+  # creating a CSIDriver object.
+  # Refer https://kubernetes-csi.github.io/docs/support-fsgroup.html for details.
+  # CSI_RBD_FSGROUPPOLICY: "None"
+
+  # (Optional) policy to modifying a volume's ownership or permissions when the CephFS PVC is being mounted.
+  # supported values are None,File,ReadWriteOnceWithFSType, if not specified nothing will be set when
```
Comment (BlaineEXE) [inline cluster/examples/kubernetes/ceph/operator.yaml:55]: The supported values documented here are likely to fall out of sync with what is supported in different k8s versions. These are in k8s 1.20, but the k8s docs say more are coming later. What if we just point them to the latest k8s docs and mention that they should pay attention to the k8s version for which ones are supported?

File: pkg/operator/ceph/csi/spec.go (hunk 1 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:616]: @travisn we do have two options here. Either the user can delete the CSI driver object and restart the operator pod. or Rook itself can detect the changed configuration and delete the CSI driver object and recreate it. which one do you prefer?

File: cluster/examples/kubernetes/ceph/operator.yaml (hunk 6 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/operator.yaml:55]: Sure I can do it.

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml (hunk 2 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:126]: I am not setting anything because Kubernetes will set the default to avoid backward compatibility issues.

File: Documentation/helm-operator.md (hunk 3 above)
Comment (Madhu-1) [reply Documentation/helm-operator.md:104]: I can send a separate PR to improve the default and available configuration for CSI.

File: cluster/charts/rook-ceph/values.yaml [hunk 7]
```diff
@@ -62,6 +62,16 @@ csi:
   # (Optional) set user created priorityclassName for csi provisioner pods.
   # provisionerPriorityClassName: system-cluster-critical
 
+  # (Optional) policy to modifying a volume's ownership or permissions when the RBD PVC is being mounted.
+  # supported values are documented at https://kubernetes-csi.github.io/docs/support-fsgroup.html
```
Comment (mykaul) [inline cluster/charts/rook-ceph/values.yaml:66]: nit: if it's user facing (vs. developer), https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods is a better link?

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml (hunk 2 above)
Comment (travisn) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:126]: What is the default behavior? I was thinking that if we set the value to be the same as the default, we wouldn't have a backward compat issue.

File: pkg/operator/ceph/csi/spec.go [hunk 8]
```diff
@@ -601,6 +611,12 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
 	}
 
 	if apierrors.IsNotFound(err) {
+		if fsGroupPolicy != "" {
+			policy := k8scsi.FSGroupPolicy(fsGroupPolicy)
+			// FSGroupPolicy field is immutable, should be set only during
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:616]: > @travisn we do have two options here. Either the user can delete the CSI driver object and restart the operator pod. or Rook itself can detect the changed configuration and delete the CSI driver object and recreate it. which one do you prefer?

I expect it will be rare enough to change the default that we can just document the need to delete it and restart the operator. Or if deleting it when needed is almost as simple as writing the doc, maybe we should just do it. I'm fine either way, you can decide.

File: cluster/charts/rook-ceph/values.yaml (hunk 7 above)
Comment (Madhu-1) [reply cluster/charts/rook-ceph/values.yaml:66]: This is more of a CSI configuration that will be done/changed(rarely) by the admin. The link I added contains the different options the admin can change which is applicable for all the users who are consuming the pvc. Users won't have the option to change values.

File: pkg/operator/ceph/csi/spec.go (hunk 8 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: Added the default now `ReadWriteOnceWithFSType` as per https://kubernetes-csi.github.io/docs/support-fsgroup.html and added logic to recreate the csidriver object if the policy is different to avoid manual intervention.

File: pkg/operator/ceph/csi/spec.go [hunk 9]
```diff
@@ -586,9 +596,31 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
...
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
 	csidrivers := clientset.StorageV1beta1().CSIDrivers()
 	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
 	if err == nil {
+		// As FSGroupPolicy field is immutable, should be set only during
+		// create time.
+		// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+		if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil {
+			if driver.Spec.FSGroupPolicy != csiDriver.Spec.FSGroupPolicy {
```
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/operator/ceph/csi/spec.go:610]: Actually we also need to check if one is nil while the other is not, which will be missed if we simply change to a dereference. What about the following replacement for lines 609-610?
```go
if driver.Spec.FSGroupPolicy != csiDriver.Spec.FSGroupPolicy {
  if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy == *csiDriver.Spec.FSGroupPolicy { 
    // the policy did not change, no need to update
    return nil
  }
  // re-create the driver
}
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:610]: Just modified it a little, if the policy changes, we will recreate the object, or else we can still try to update the CSIDriver object. because sometimes we might need to update the Mutable objects in the future.

File: pkg/operator/ceph/csi/spec.go [hunk 10]
```diff
@@ -486,13 +487,23 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
...
 		if err != nil {
 			return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
 		}
 	}
 	if EnableCephFS {
-		err = createCSIDriverInfo(ctx, clientset, CephFSDriverName)
+		fsGroupPolicyForCephFS, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_CEPHFS_FSGROUPPOLICY", "")
+		if err != nil {
+			// logging a warning and intentionally continuing with the default log level
+			logger.Warningf("failed to load CSI_CEPHFS_FSGROUPPOLICY %v", err)
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:504]:
```suggestion
			logger.Warningf("failed to load CSI_CEPHFS_FSGROUPPOLICY. %v", err)
```

File: pkg/operator/ceph/csi/spec.go [hunk 11]
```diff
@@ -610,6 +632,19 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
 	return err
 }
 
+func reCreateCSIDriverInfo(ctx context.Context, csiClient v1beta1.CSIDriverInterface, csiDriver *k8scsi.CSIDriver) error {
+	err := csiClient.Delete(ctx, csiDriver.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return err
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:638]: what about wrapping the error like we commonly do?

**Review by leseb — CHANGES_REQUESTED** [review]
nit

**Review by leseb — APPROVED** [review]

File: pkg/operator/ceph/csi/spec.go [hunk 12]
```diff
@@ -586,9 +597,20 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
...
+		policy := k8scsi.FSGroupPolicy(fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
 	csidrivers := clientset.StorageV1beta1().CSIDrivers()
 	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
 	if err == nil {
+		// As FSGroupPolicy field is immutable, should be set only during
+		// create time.
+		// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+		if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:610]: If they are enabling the setting for the first time (changing the policy from blank to some value), or disabling the setting (changing the policy back to blank), will the driver be recreated? It seems like it will only be updated if the value is already set and is being changed.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:610]: It supports only 3 values as per the doc I linked above, setting empty value is not allowed. If the kubernetes is upgraded from 1.18 to 1.19 the csidriver value will be set to default as per my testing (I can test the upgrade again and see)

Comment (travisn) [reply pkg/operator/ceph/csi/spec.go:610]: On line 600 above it looks like the policy could be blank. That would at least happen on upgrades, right? So if they upgraded and then applied the new policy, the csi driver wouldn't be restarted?

Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:610]: After upgrade kubernetes will set the value to ReadWriteOnceWithFSType which is default. The check is for safer side and for kubernetes versions less than 1.19, even if we set the fsgrouppolicy it will be discarded.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:610]: User is not suppose to set it to empty in configmap ,if he set it to empty the object update will fail. Its admin responsibility to set it to valid options.

Comment (travisn) [reply pkg/operator/ceph/csi/spec.go:610]: Where will it fail if the value is not set in the configmap? It seems like it just won't update the driver in that case. 

The Rook upgrade does not require adding the values to the configmap so the settings may not be in the configmap at all. 

The default values (the last param) to load the settings on lines 490 and 501 are blank. Do we just need to set those to the expected default value? Then we will guarantee it's always set.

File: pkg/operator/ceph/csi/spec.go [hunk 13]
```diff
@@ -486,13 +487,23 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
 	}
 
 	if EnableRBD {
-		err = createCSIDriverInfo(ctx, clientset, RBDDriverName)
+		fsGroupPolicyForRBD, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_RBD_FSGROUPPOLICY", "")
+		if err != nil {
+			// logging a warning and intentionally continuing with the default log level
+			logger.Warningf("failed to load CSI_RBD_FSGROUPPOLICY. %v", err)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:493]: I think it would be good to indicate here what users can expect from the operator in this case. The operator continues, but what setting will it use? Will it disable the fsgrouppolicy?

File: pkg/operator/ceph/csi/spec.go [hunk 14]
```diff
@@ -486,13 +487,23 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
...
 		if err != nil {
 			return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
 		}
 	}
 	if EnableCephFS {
-		err = createCSIDriverInfo(ctx, clientset, CephFSDriverName)
+		fsGroupPolicyForCephFS, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_CEPHFS_FSGROUPPOLICY", "")
+		if err != nil {
+			// logging a warning and intentionally continuing with the default log level
+			logger.Warningf("failed to load CSI_CEPHFS_FSGROUPPOLICY. %v", err)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:504]: Ditto here. What expectations can the users have from the operator's config moving forward?

File: pkg/operator/ceph/csi/spec.go [hunk 15]
```diff
@@ -586,9 +597,20 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
...
 	}
+	if fsGroupPolicy != "" {
+		policy := k8scsi.FSGroupPolicy(fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
 	csidrivers := clientset.StorageV1beta1().CSIDrivers()
 	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
 	if err == nil {
+		// As FSGroupPolicy field is immutable, should be set only during
+		// create time.
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:608]: Nit: move these 2 words to the previous line?

File: pkg/operator/ceph/csi/spec.go [hunk 16]
```diff
@@ -610,6 +632,20 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
...
 
+func reCreateCSIDriverInfo(ctx context.Context, csiClient v1beta1.CSIDriverInterface, csiDriver *k8scsi.CSIDriver) error {
+	err := csiClient.Delete(ctx, csiDriver.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete CSIDriver object for driver %q", csiDriver.Name)
+	}
+	logger.Infof("CSIDriver object deleted for driver %q", csiDriver.Name)
+	_, err = csiClient.Create(ctx, csiDriver, metav1.CreateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to create CSIDriver object for driver %q", csiDriver.Name)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:643]:
```suggestion
		return errors.Wrapf(err, "failed to recreate CSIDriver object for driver %q", csiDriver.Name)
```

File: pkg/operator/ceph/csi/spec.go [hunk 17]
```diff
@@ -610,6 +632,20 @@ func createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, na
...
+	}
+	logger.Infof("CSIDriver object created for driver %q", csiDriver.Name)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go:645]:
```suggestion
	logger.Infof("CSIDriver object recreated for driver %q", csiDriver.Name)
```

File: Documentation/helm-operator.md [hunk 18]
```diff
@@ -101,7 +101,9 @@ The following tables lists the configurable parameters of the rook-operator char
...
+| `csi.rbdFSGroupPolicy`             | Policy to modifying a volume's ownership or permissions when the RBD PVC is being mounted                                      | ReadWriteOnceWithFSType                                   |
```
Comment (BlaineEXE) [inline Documentation/helm-operator.md:105]:
```suggestion
| `csi.rbdFSGroupPolicy`             | Policy for modifying a volume's ownership or permissions when the RBD PVC is being mounted                                      | ReadWriteOnceWithFSType                                   |
```

File: Documentation/helm-operator.md [hunk 19]
```diff
@@ -101,7 +101,9 @@ The following tables lists the configurable parameters of the rook-operator char
...
+| `csi.cephFSFSGroupPolicy`          | Policy to modifying a volume's ownership or permissions when the CephFS PVC is being mounted                                   | ReadWriteOnceWithFSType                                   |
```
Comment (BlaineEXE) [inline Documentation/helm-operator.md:106]:
```suggestion
| `csi.cephFSFSGroupPolicy`          | Policy for modifying a volume's ownership or permissions when the CephFS PVC is being mounted                                   | ReadWriteOnceWithFSType                                   |
```

File: Documentation/helm-operator.md (hunk 3 above)
Comment (BlaineEXE) [reply Documentation/helm-operator.md:104]: Is there a link to explain what omap does for users?

Also, does omap need to be capitalized (i.e., OMAP)?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 12 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:630]: added the default value now and also logging what Rook is defaulting.

**Comment by Madhu-1** [comment]
@BlaineEXE we are yet to add omap documented which will come in Async DR doc. The omap table got adjusted due to my IDE. correct it now. This PR won't touch anything related to omap.

**Review by travisn — APPROVED** [review]

**Review by leseb — APPROVED** [review]

**Comment by Madhu-1** [comment]
@BlaineEXE can you please review this one.

**Review by BlaineEXE — APPROVED** [review]
